runtime/serde/serde-json/common/src/aws/smithy/kotlin/runtime/serde/json/JsonLexer.kt

```diff
@@ -0,0 +1,433 @@
+            '"' -> JsonToken.String(readQuoted())
+            't', 'f', 'n' -> readKeyword()
+            '-', in '0'..'9' -> readNumber()
+            null -> JsonToken.EndDocument
+            else -> unexpectedToken(chr, "{", "[", "\"", "null", "true", "false", "<number>")
+        }
+
+    /**
+     * Read based on the number spec : https://www.json.org/json-en.html
+     * [-]0-9[.[0-9]][[E|e][+|-]0-9]
+     */
+    private fun readNumber(): JsonToken {
+        val value = buildString {
+            if (peekChar() == '-') {
+                append(nextOrThrow())
+            }
+            readDigits(this)
+            if (peekChar() == '.') {
+                append(nextOrThrow())
+                readDigits(this)
+            }
+            if (peekChar() in EXP) {
+                append(nextOrThrow())
+                if (peekChar() in PLUS_MINUS) {
+                    append(nextOrThrow())
+                }
+                readDigits(this)
+            }
+        }
+        lexerCheck(value.isNotEmpty()) { "Invalid number, expected '-' || 0..9, found ${peekChar()}" }
+        return JsonToken.Number(value)
+    }
+
+    private fun readDigits(appendable: Appendable) {
+        while (peekChar() in DIGITS) {
+            appendable.append(nextOrThrow())
+        }
+    }
+
+    /**
+     * Read a quoted JSON string out of the stream
+     */
+    private fun readQuoted(): String {
+        consume('"')
+        // read bytes until a non-escaped end-quote
+        val start = idx
+        var chr = peekOrThrow()
+        var needsUnescaped = false
+        while (chr != '"') {
+            // handle escapes
+            when (chr) {
+                '\\' -> {
+                    needsUnescaped = true
+                    // consume escape backslash
+                    nextOrThrow()
+                    when (val byte = nextOrThrow()) {
+                        'u' -> {
+                            if (idx + 4 >= data.size) throw DeserializationException("Unexpected EOF")
+                            idx += 4
+                        }
+                        '\\', '/', '"', 'b', 'f', 'r', 'n', 't' -> { } // already consumed
+                        else -> fail("Invalid escape character: `$byte`", idx - 1)
+                    }
+                }
+                else -> {
+                    if (chr.isControl()) fail("Unexpected control character: `$chr`")
+                    idx++
+                }
+            }
+
+            chr = peekOrThrow()
+        }
+
+        val value = data.decodeToString(start, idx)
+        consume('"')
+        return if (needsUnescaped) value.unescape() else value
+    }
+
+    private fun readKeyword(): JsonToken = when (val ch = peekOrThrow()) {
+        't' -> readLiteral("true", JsonToken.Bool(true))
+        'f' -> readLiteral("false", JsonToken.Bool(false))
+        'n' -> readLiteral("null", JsonToken.Null)
+        else -> fail("Unable to handle keyword starting with '$ch'")
```

Comment:
**Question**: This doesn't give the user a lot to go on. Can we give the full keyword we failed to parse?

Reply:
well it gives us the position and the character. We don't know what the word is since it wouldn't be a keyword. Perhaps updating the error message to `Invalid literal starting with $ch`?

Reply:
I think the starting character doesn't tell us much. You're right that position will be helpful (all the moreso if we can augment it to be line:pos and/or include the path) but the starting character doesn't help with searching or getting more context for debugging. Ideally, I'd like to see something like:

```json
{
  "a": true,
  "b": false,
  "c": trueish
}
```

Lead to the error message: `"Expected one of ['true', 'false', 'null'] but found 'trueish'"`. I believe that'd be the most helpful message we could give debuggers.

Reply:
If you don't know what you're parsing how do you know what to parse and when to stop? At this point in lexical analysis we have an invalid JSON document, it could be anything that follows.

Reply:
Yeah, once the document's known to be invalid then it gets tricky for sure.

I'm just hoping we can gobble up the text we find until we encounter a whitespace or punctuation. A regex like `"[-_a-zA-Z0-9]+"` encapsulates something that would get us the next literal word even though the document is invalid and we're going to fail parsing. If you're worried about length (i.e., potentially grabbing the remainder of some very long document) we could even cap the size of the word we'll attempt to include in the error message (e.g., no more than 20 chars).

I think this would improve debuggability if it's possible. If it's not or introduces other concerns then we'll just have to rely on whatever positioning/pathing we can get.

Reply:
We actually don't ever enter this else clause because we only get called when we match `t`, `f,` or `n`. The error we see when trying to read:

* `trueish` -> we will consume `true` just fine and then we'll see unexpected token `i` with offset info
* `ture` -> `Unexpected JSON token at offset 1; Unexpected char 'u' expected 'r'`